pkg/content/provider/oci/oci.go

```diff
@@ -0,0 +1,142 @@
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/mutate"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	"sigs.k8s.io/kustomize/kyaml/filesys"
+)
+
+type ociProvider struct {
+	image v1.Image
+	dir   string
+	fsys  filesys.FileSystem
+}
+
+var _ extensions.FileSystemProvider = &ociProvider{}
+
+func Open(ref location.Oci, options ...remote.Option) (content.Content, location.ReferenceLock, error) {
+	return open(ref.Image, ref.Directory, ref, options...)
+}
+
+func OpenLock(ref location.OciLock, options ...remote.Option) (content.Content, location.ReferenceLock, error) {
+	return open(ref.Digest, ref.Directory, ref, options...)
+}
+
+func open(name name.Reference, dir string, ref location.Reference, options ...remote.Option) (content.Content, location.ReferenceLock, error) {
+	image, err := remote.Image(name, options...)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Determine the digest of the image that was extracted
+	imageDigestHash, err := image.Digest()
+	if err != nil {
+		return nil, nil, err
+		// return nil, errors.E(op, fmt.Errorf("error calculating image digest: %w", err))
+	}
+
+	lock, err := locationmutate.Lock(ref, "sha256:"+imageDigestHash.Hex)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return &ociProvider{
+		image: image,
+		dir:   dir,
+	}, lock, nil
+}
+
+func (p *ociProvider) Close() error {
+	return nil
+}
+
+func (p *ociProvider) ProvideFileSystem() (filesys.FileSystem, string, error) {
+	if p.fsys != nil {
+		return p.fsys, "/", nil
+	}
+
+	fsys := filesys.MakeFsInMemory()
+
+	if err := pullAndExtract(p.image, p.dir, fsys, "/"); err != nil {
+		return nil, "", err
+	}
+
+	p.fsys = fsys
+	return fsys, "/", nil
+}
+
+// pullAndExtract uses current credentials (gcloud auth) to pull and
+// extract (untar) image files to target directory. The desired version or digest must
+// be in the imageName, and the resolved image sha256 digest is returned.
+func pullAndExtract(image v1.Image, srcDir string, fsys filesys.FileSystem, dstDir string) error {
+	// const op errors.Op = "oci.pullAndExtract"
+
+	// Stream image files as if single tar (merged layers)
+	ioReader := mutate.Extract(image)
+	defer ioReader.Close()
+
+	// Write contents to target dir
+	// TODO look for a more robust example of an untar loop
+	tarReader := tar.NewReader(ioReader)
+	for {
+		hdr, err := tarReader.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		relPath, err := filepath.Rel(srcDir, hdr.Name)
+		if err != nil || strings.HasPrefix(relPath, "..") {
+			// skip because hdr.Name not contained under srcDir
+			continue
+		}
+
+		path := filepath.Join(dstDir, relPath)
+		switch {
+		case hdr.FileInfo().IsDir():
+			if err := fsys.MkdirAll(path); err != nil {
+				return err
+			}
+		default:
+			file, err := fsys.Create(path)
+			if err != nil {
+				return err
+			}
+			defer file.Close()
+
+			_, err = io.Copy(file, tarReader)
```

Comment:
This project https://github.com/mholt/archiver is an interesting one that provides `fs.FS` from an archive. Probably works better with streams.

Reply:
Indeed - and the `/pkg/content/provider/*/*.go` are exactly where those kind of choices and changes can be made in a place that's dedicated only to the task of exposing one type of location as content.

Doesn't need to understand argument parsing, package structures, or file formats... Also being mainly lowercase doesn't need to worry about becoming public surface area that will cause breaking changes if updated.

And also to @martinmaly question on related files, can be replaced if someone gives you a custom ContentOpener as an open.Option... e.g. in places where the kptlib is used that shouldn't exec git to clone to tmp folder --- all you need is to intercept the moment where the default &gitProvider{} would have been returned with your own --- and everything else in the system runs as they would because they don't care that's a difference